nce/nce_pairwise_mp/trainers/qa_trainer.py

```diff
@@ -0,0 +1,220 @@
+
+            del features
+            # pack the selected pos and neg samples into the torchtext batch and train
+            if epoch != 1:
+                true_batch_size = len(new_train_neg["answer"])
+                if true_batch_size != 0:
+                    for j in range(true_batch_size):
+                        new_train_neg["answer"][j] = F.pad(new_train_neg["answer"][j],
+                                                           (0, max_len_a - new_train_neg["answer"][j].size()[0]),
+                                                           value=1)
+                        new_train_pos["answer"][j] = F.pad(new_train_pos["answer"][j],
+                                                           (0, max_len_a - new_train_pos["answer"][j].size()[0]),
+                                                           value=1)
+                        new_train_pos["question"][j] = F.pad(new_train_pos["question"][j],
+                                                             (0, max_len_q - new_train_pos["question"][j].size()[0]),
+                                                             value=1)
+                        new_train_neg["question"][j] = F.pad(new_train_neg["question"][j],
+                                                             (0, max_len_q - new_train_neg["question"][j].size()[0]),
+                                                             value=1)
+
+
+                    pos_batch = get_batch(new_train_pos["question"], new_train_pos["answer"], new_train_pos["ext_feat"],
+                                          true_batch_size)
+                    neg_batch = get_batch(new_train_neg["question"], new_train_neg["answer"], new_train_neg["ext_feat"],
+                                          true_batch_size)
+
+                    self.model.train()
+                    self.optimizer.zero_grad()
+                    output = self.model([pos_batch, neg_batch])
+
+                    cmp = output[:, 0] > output[:, 1]
+                    acc += sum(cmp.data.cpu().numpy())
+                    tot += true_batch_size
+
+                    loss = self.loss(output[:, 0], output[:, 1], self.margin_label)
+                    if len(new_near_score) != 0 and self.weighting:
+                        # Element wise weighting hasn't been implemented for MarginRankingLoss by pytorch
+                        # Apply the mean score and set branch size to 1 as weighting implementation temporarily here
+                        # Reference: https://github.com/pytorch/pytorch/issues/264
+                        # loss *= torch.autograd.variable.Variable(torch.from_numpy(np.array(new_near_score)).cuda())
+                        loss *= np.mean(new_near_score)
+
+                    loss_num = loss.data.cpu().numpy()[0]
+                    total_loss += loss_num
+                    loss.backward()
+                    self.optimizer.step()
+
+                    del output
```

Comment:
does this help with forcing Python to release memory?

Reply:
No in fact. In my case adding it does not solve the Out Of Memory exception but does not affect the result. I should remove it. THX!